### PR TITLE
feat: integrate starknet-devnet as part of the project to make it easy to play around

### LINKED ISSUE #132: feat: run starknet-devnet in the background when not found.
# Feature Request

When starting the RPC server, it will be interesting if there is no need to run the starknet-devnet and deploy Kakarot when the **STARKNET_NETWORK == "devnet"**.

This has already been done for the sandbox [here](https://github.com/sayajin-labs/kakarot-rpc-sandbox/commit/d781f95fa1cda4031140b241e82b0041f9777e59).

The above feature makes it a smooth and fast experience for the end-user.

**Describe Preferred Solution**

We can run this in a subprocess and stop it when our RPC service is stopped.

**If the feature request is approved, would you be willing to submit a PR?**
Yes

### LINKED ISSUE #133: bug: source `.env.example` fails
# Bug Report

**Kakarot RPC version:**

**HEAD** ->  [b4e148fc72c5a414199fe4bec3d0e3c3726d5fee](https://github.com/sayajin-labs/kakarot-rpc/commit/b4e148fc72c5a414199fe4bec3d0e3c3726d5fee)

**Current behavior:**

```shell
cp .env.example .env && source .env
```

The above will result in the following error:
```shell
source .env.example
.env.example:2: command not found: STARKNET_RPC_URL
.env.example:3: command not found: RUST_LOG
.env.example:4: command not found: KAKAROT_HTTP_RPC_ADDRESS
.env.example:5: command not found: KAKAROT_ADDRESS
.env.example:6: command not found: PROXY_ACCOUNT_CLASS_HASH
```

**Expected behavior:**

The `.env.example` should be in the correct format, so that when copied into `.env` and then used with `source`,  it should work fine.

**Steps to reproduce:**

```shell
cp .env.example .env && source .env
```

### PR DESCRIPTION
**Pull request by bajpai244**
As a user, someone might want to use it with devnet for fast prototyping, but they would have to deploy the contracts again, and again.

### TL;DR:

This PR adds starknet-devnet as a dependency and allows running it via a simple `make devnet`, where the running devnet already has contracts deployed so that a user can directly start playing around.

So, for an end-user, using this in dev mode, all they have to do is:

Run devnet: 
```shell
make devnet
```

Run RPC:
```shell
make run
```

Here is a description of the changes introduced via this PR:

### Virtual Environment with Poetry

Poetry has been used to manage a virtual environment to have starknet-devnet installed, without messing around with the existing system configuration of the user!

### Makefile

A Unified makefile for doing all things inside the RPC:
- tests
- build release
- run devnet
- run RPC in dev mode
- install dependencies, etc.

This gives a very easy-to-consume experience for the final user, whatever they do want to around the project, they can simply use make!

### Deployments directory

This directory is handy, as it contains information related various deployments of Kakarot, which could be useful for someone using the RPC. { borrowed from  [rpc-sandbox](https://github.com/sayajin-labs/kakarot-rpc-sandbox) } 

### `.env.example`

The formatting of `.env.example` was wrong as mentioned in #133 , this PR also fixes that + it populates this with default values, which when copied into a `.env` will work out of the box when in dev mode with `starknet-devnet` { given user is running it via `make devnet` }  

### README

The readme has been updated as per these new changes.

### Why are we not automatically running devnet in background as described in the original issue?

- **too much magic**: Automatically running it in the background can be too much magic, if a user is using custom Kakarot contracts, and forgets to run their devnet, the RPC will automatically spinup one, and the user might be confused for some time that why their changes are not reflecting.
- **require more code added to RPC**: This also requires more code to be added to the RPC! Which means more tests and more maintenance around it.

The approach in this PR achieves the goal of making it easy for user to play around the system { i.e not deploy contracts again and again }, without making any change to the source code.

Time spent on this PR: 4-5 hours

Closes #133 
Closes #132 

# Pull Request type

Please check the type of change your PR introduces:

- [X] Bugfix
- [X] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no API changes)
- [ ] Build-related changes
- [ ] Documentation content changes
- [ ] Testing
- [ ] Other (please describe)

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+# run devnet
+devnet: poetry-install 
+	 poetry run starknet-devnet --seed 0 --disable-rpc-request-validation --load-path deployments/devnet/devnet.pkl --timeout 300
+
+# build
+build:
+	cargo build --all --release
+
+# run
+run:
```
Comment (greged93) [inline Makefile:19]: it looks like given the current format for the `example.env` file, sourcing won't work. I would update the `example.env` and remove all spaces.
```
# Copy this file to .env and fill in the values
STARKNET_RPC_URL=http://0.0.0.0:5050/rpc
RUST_LOG=debug
KAKAROT_HTTP_RPC_ADDRESS=0.0.0.0:3030
KAKAROT_ADDRESS=0x1234
PROXY_ACCOUNT_CLASS_HASH=0x1234
```

File: pyproject.toml [hunk 2]
```diff
@@ -0,0 +1,16 @@
+[tool.poetry]
+name = "kakarot-rkakarot-rpc"
```
Comment (greged93) [inline pyproject.toml:2]: Seems like a typo

File: deployments/devnet/declarations.json [hunk 3]
```diff
@@ -0,0 +1,7 @@
+{
```
Comment (greged93) [inline deployments/devnet/declarations.json:1]: I understand that you need the `devnet.pkl` but are you using all the other deployment and declaration files? Otherwise suggest to move the kakarot_address and proxy_account_class_hash to `example.env` and be rid with the rest.

**Review by greged93 — COMMENTED** [review]
Some small comments to fix

File: Makefile (hunk 1 above)
Comment (bajpai244) [reply Makefile:19]: Oh! I already had an issue opened for it! Looks like I forgot to push the commit to fix it!

-> https://github.com/sayajin-labs/kakarot-rpc/issues/133

File: pyproject.toml (hunk 2 above)
Comment (bajpai244) [reply pyproject.toml]: fixed

File: deployments/devnet/declarations.json (hunk 3 above)
Comment (bajpai244) [reply deployments/devnet/declarations.json:1]: I think the other deployment files can be of good help if someone wants to run the RPC over the test net!

They wouldn't have to come and ping us on telegram to ask for these addresses and class hash.

But if you think removing them seems better then I can do so! Lemme know your views on it!

Comment (greged93) [reply deployments/devnet/declarations.json:1]: I think the issue is that the [CI](https://github.com/sayajin-labs/kakarot/blob/main/.github/workflows/deployments.yml) from kakarot will re-deploy all contracts every time we push on main. Meaning these addresses could be out of date as soon as a PR is merged, leading users to use incorrect addresses.
We could add this to the README and point users to the [deployments](https://github.com/sayajin-labs/kakarot/actions/workflows/deployments.yml) on kakarot repo.

Comment (bajpai244) [reply deployments/devnet/declarations.json]: Makes sense!

Got this done -> [link](https://github.com/sayajin-labs/kakarot-rpc/pull/134/commits/1b6451b812f00c16e54fbee491899e16dda350ab)

**Review by Eikix — APPROVED** [review]
Thanks a lot for your energy @bajpai244 🔥

File: Makefile [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+install:
+	sh kakarotup
+
+# install dependencies, automatically creates a virtual environment
+poetry-install:
+	poetry install
+
+# run devnet
+devnet: poetry-install 
+	 poetry run starknet-devnet --seed 0 --disable-rpc-request-validation --load-path deployments/devnet.pkl --timeout 300
```
Comment (Eikix) [inline Makefile:12]: As per Ivan from Shardlabs, maybe it's better to increase timeout to 5s

> Hello! You may experience timeouts in the devnet (during declare). This would potentially fix the issue https://github.com/0xSpaceShard/starknet-devnet/issues/434 ...Until then, the workaround is to start devnet with a higher timeout, for example --timeout 5000 !

File: Makefile [hunk 5]
```diff
@@ -0,0 +1,29 @@
+#install kakarot
+
+install:
+	sh kakarotup
```
Comment (Eikix) [inline Makefile:4]: Let me check, but I've received some bug report that sh kakarotup does not work, do you know if this is true? I'll check locally

**Comment by Eikix** [comment]
Will merge after those two comments. 
PS: Created #138 so you can leave kakarotup as is, it'll be replaced in a subsequent PR imho.

File: Makefile (hunk 4 above)
Comment (bajpai244) [reply Makefile]: done -> [link](https://github.com/sayajin-labs/kakarot-rpc/pull/134/commits/afa06f24336adbf34b6706bbc1ff488aa76eb556)

File: Makefile (hunk 5 above)
Comment (bajpai244) [reply Makefile:4]: I guess we can address this in a later PR as you mentioned below!

**Review by Eikix — APPROVED** [review]
lgtm 🔥

**Comment by bajpai244** [comment]
lgtm 🔥
